File: accounts-daemon/token-provisioning/passwordprovisioningmethod.cpp
//
// Created by victor on 20/09/24.
//

#include "passwordprovisioningmethod.h"

#include "dbus/accountmanager.h"
#include "dbus/twofactor.h"
#include "dbus/user.h"
#include "dbus/useraccount.h"

#include <QDateTime>
#include <QSqlQuery>

struct PasswordProvisioningMethodPrivate {
};

PasswordProvisioningMethod::PasswordProvisioningMethod(AccountManager* parent) :
    TokenProvisioningMethod(parent) {
}

QString PasswordProvisioningMethod::tokenProvisioningMethod() const {
    return QStringLiteral("password");
}

TokenProvisioningMethod::ProvisionResult PasswordProvisioningMethod::provision(QVariantMap options, TokenProvisioningManager::TokenProvisioningPurpose provisioningPurpose) const {
    const auto username = options.value("username").toString();
    auto password = options.value("password").toString();

    if (username.isEmpty() || password.isEmpty()) {
        return {0, Utils::InvalidInput};
    }

    const auto id = accountManager()->userIdByUsername(username);
    if (id == 0) {
        return {0, Utils::NoAccount};
    }

    QSqlQuery userQuery;
    userQuery.prepare("SELECT * FROM users WHERE id=:id");
    userQuery.bindValue(":id", id);
    userQuery.exec();
    userQuery.next();

    // Ensure the password is correct
    const auto passwordHash = userQuery.value("password").toString();
    if (passwordHash.startsWith("!")) {
        return {0, Utils::DisabledAccount};
    }

    auto* account = UserAccount::accountForId(id);

    // Now check for password resets
    bool havePasswordReset = false;
    {
        QSqlQuery resetQuery;
        resetQuery.prepare("SELECT * FROM passwordresets WHERE userid=:id");
        resetQuery.bindValue(":id", id);
        resetQuery.exec();
        while (resetQuery.next()) {
            if (resetQuery.value("expiry").toLongLong() > QDateTime::currentMSecsSinceEpoch()) {
                havePasswordReset = true;
                const auto temporaryPassword = resetQuery.value("temporarypassword").toString();
                if (Utils::verifyHashedPassword(password, temporaryPassword)) {
                    if (!options.contains("newPassword")) {
                        return {0, Utils::PasswordResetRequired};
                    }

                    // Set the new password on this user account
                    const auto newPassword = options.value("newPassword").toString();
                    if (const auto error = account->user()->setPassword(newPassword)) {
                        return {0, error};
                    }

                    QSqlQuery deleteQuery;
                    deleteQuery.prepare("DELETE FROM passwordresets WHERE userid=:id");
                    deleteQuery.bindValue(":id", id);
                    deleteQuery.exec();

                    password = newPassword;
                }
            }
        }
    }

    if (passwordHash == "x") {
        // Check if there is already a pending password reset.
        // If there is already a pending password reset, tell the user that their password is incorrect instead.
        if (havePasswordReset) {
            return {0, Utils::IncorrectPassword};
        }

        return {0, Utils::PasswordResetRequired};
    }

    if (!Utils::verifyHashedPassword(password, passwordHash)) {
        return {0, Utils::IncorrectPassword};
    }

    // Check TOTP if we're doing this to log in
    if (provisioningPurpose == TokenProvisioningManager::TokenProvisioningPurpose::LoginToken) {
        QSqlQuery otpQuery;
        otpQuery.prepare("SELECT * FROM otp WHERE userid=:id");
        otpQuery.bindValue(":id", id);
        if (!otpQuery.exec()) {
            return {0, Utils::QueryError};
        }

        if (otpQuery.next()) {
            if (otpQuery.value("enabled").toBool()) {
                if (!options.contains("otpToken")) {
                    return {0, Utils::TwoFactorRequired};
                }

                const auto otpSecret = otpQuery.value("otpkey").toString();
                const auto otpKey = options.value("otpToken").toString();
                if (!Utils::isValidOtpKey(otpKey, otpSecret)) {
                    // Check the backup keys
                    QSqlQuery backupsQuery;
                    backupsQuery.prepare("SELECT * FROM otpbackup WHERE userid=:id");
                    backupsQuery.bindValue(":id", id);
                    backupsQuery.exec();

                    bool foundValidBackupKey = false;
                    while (backupsQuery.next()) {
                        auto backupKey = backupsQuery.value("backupkey").toString();
                        if (!backupsQuery.value("used").toBool() && backupKey == otpKey) {
                            QSqlQuery updateBackupQuery;
                            updateBackupQuery.prepare("UPDATE otpbackup SET used=:used WHERE userid=:id AND backupkey=:key");
                            updateBackupQuery.bindValue(":used", true);
                            updateBackupQuery.bindValue(":id", id);
                            updateBackupQuery.bindValue(":key", backupKey);
                            if (!updateBackupQuery.exec()) {
                                return {0, Utils::QueryError};
                            }

                            account->twoFactor()->reloadBackupKeys();

                            foundValidBackupKey = true;
                        }
                    }

                    if (!foundValidBackupKey) {
                        return {0, Utils::TwoFactorRequired};
                    }
                }
            }
        }
    }

    return {id, Utils::NoError};
}

bool PasswordProvisioningMethod::available(quint64 userId, QString application, TokenProvisioningManager::TokenProvisioningPurpose provisioningPurpose) const {
    // This method is always available
    // TODO: For account modification this method should not be available
    return true;
}

File: src/token_provisioning/password_provisioning_method.rs
use crate::accounts_manager::user_id_by_username;
use crate::error::Error;
use crate::token_provisioning::{ProvisionResult, TokenProvisioningPurpose};
use crate::{VariantMap, is_valid_otp_key, verify_hashed_password};
use sqlx::{PgPool, Row};
use zbus::Connection;
use zvariant::Value;
use crate::bus::user_object;

pub async fn provision(
    bus: &Connection,
    database: &PgPool,
    options: VariantMap<'_>,
    purpose: TokenProvisioningPurpose,
) -> Result<ProvisionResult<'static>, Error> {
    let Some(username) = options.get("username").and_then(|username| match username {
        Value::Str(s) => Some(s.to_string()),
        _ => None,
    }) else {
        return Err(Error::InvalidInput);
    };

    let Some(mut password) = options.get("password").and_then(|password| match password {
        Value::Str(s) => Some(s.to_string()),
        _ => None,
    }) else {
        return Err(Error::InvalidInput);
    };

    let id = user_id_by_username(database, &username).await?;
    let user_record = sqlx::query("SELECT * FROM users WHERE id = $1")
        .bind(id)
        .fetch_one(database)
        .await?;

    let password_hash = user_record.try_get::<String, _>("password")?;
    if password_hash.starts_with("!") {
        return Err(Error::DisabledAccount);
    }

    // Now check for password resets
    let have_password_reset = {
        let mut have_password_reset = false;
        let resets = sqlx::query("SELECT * FROM passwordresets WHERE userid = $1")
            .bind(id)
            .fetch_all(database)
            .await?;
        for reset in resets {
            if reset.try_get::<i64, _>("expiry")? <= chrono::Utc::now().timestamp() {
                continue;
            }
            have_password_reset = true;
            let temporary_password = reset.try_get::<String, _>("temporarypassword")?;
            if verify_hashed_password(&password, &temporary_password) {
                let Some(new_password) =
                    options
                        .get("newPassword")
                        .and_then(|new_password| match new_password {
                            Value::Str(s) => Some(s.to_string()),
                            _ => None,
                        })
                else {
                    return Err(Error::PasswordResetRequired);
                };

                user_object(bus, database, id, async |account| -> Result<(), Error> {
                    account.set_password(&new_password).await?;
                    Ok(())
                }).await??;

                // The password has been reset, so delete all the password resets for this user
                let _ = sqlx::query("DELETE FROM passwordresets WHERE userid = $1")
                    .bind(id)
                    .execute(database)
                    .await;

                password = new_password;
            }
        }
        have_password_reset
    };

    if password_hash == "x" {
        // Check if there is already a pending password reset.
        // If there is already a pending password reset, tell the user that their password is incorrect instead.
        return Err(if have_password_reset {
            Error::IncorrectPassword
        } else {
            Error::PasswordResetRequestRequired
        });
    }

    if !verify_hashed_password(&password, &password_hash) {
        return Err(Error::IncorrectPassword);
    }

    // Check TOTP if we're doing this to log in
    if purpose == TokenProvisioningPurpose::Login {
        match sqlx::query("SELECT * FROM otp WHERE userId = $1")
            .bind(id)
            .fetch_one(database)
            .await
        {
            Ok(row) => {
                if row.try_get::<bool, _>("enabled")? {
                    let Some(otp_token) =
                        options
                            .get("otpToken")
                            .and_then(|otp_token| match otp_token {
                                Value::Str(s) => Some(s.to_string()),
                                _ => None,
                            })
                    else {
                        return Err(Error::TwoFactorRequired);
                    };

                    let otp_secret = row.try_get::<String, _>("otpkey")?;
                    if !is_valid_otp_key(&otp_token, &otp_secret) {
                        // Check the backup keys
                        let backup_keys = sqlx::query("SELECT * FROM otpbackup WHERE userid = $1")
                            .bind(id)
                            .fetch_all(database)
                            .await?;
                        let Some(valid_backup_key) = backup_keys.iter().find_map(|key| {
                            if key.try_get::<bool, _>("used").ok()? {
                                None
                            } else {
                                let backup_key = key.try_get::<String, _>("backupkey").ok()?;
                                if backup_key == otp_token {
                                    Some(backup_key)
                                } else {
                                    None
                                }
                            }
                        }) else {
                            return Err(Error::TwoFactorRequired)
                        };

                        // Mark the valid backup key as used
                        sqlx::query("UPDATE otpbackup SET used = true WHERE backupkey = $1 AND userid = $2")
                            .bind(valid_backup_key)
                            .bind(id)
                            .execute(database)
                            .await?;
                    }
                }
            }
            Err(sqlx::Error::RowNotFound) => {
                // No TOTP required to log in
            }
            Err(e) => return Err(e.into()),
        }
    }

    Ok(ProvisionResult::Success(id))
}
